File: chapter4/list/list.cpp
#include "list.h"

template <class T>
struct Node
{
  Node* pred;
  Node* succ;
  T value;
};

////////////// LIST CLASS DEFINITION //////////////

template <class T>
List<T>::List()
{
  Node<T>* n = new Node<T>();
  n->pred = n;
  n->succ = n;
  first_node = n;
}
/*
template <class T>
List<T>::List(const List<T>& li)
{
  for (iterator it=li.begin(); it!=li.end(); it++)
  {
    insert_before(end(),*it);
  }
  }*/

template <class T>
bool List<T>::empty()
{
  return ((this->first_node->pred == this->first_node->succ) &&
          (this->first_node->succ == this->first_node));
}

template <class T>
List_iterator<T> List<T>::begin()
{
  return List_iterator<T>(first_node->succ);
}

template <class T>
List_iterator<T> List<T>::end()
{
  //return first invalid value (so that it allows looping using != condition)
  return List_iterator<T>(first_node);
}

template <class T>
bool List<T>::finished(iterator p)
{
  return (p.node == first_node);
}

template <class T>
void List<T>::write(iterator p, T v)
{
  p.node->value = v;
}

template <class T>
void List<T>::insert_before(iterator p, T v)
{
  Node<T>* t = new Node<T>();
  t->value = v;
  t->pred = p.node->pred;
  t->pred->succ = t;
  t->succ = p.node;
  p.node->pred = t;
}

//the iterator is passed by reference because after the deletion it is moved
//to the next element
template <class T>
void List<T>::remove(iterator& p)
{
  Node<T>* t;
  p.node->pred->succ = p.node->succ;
  p.node->succ->pred = p.node->pred;
  t = p.node;
  p++;
  delete t;
}


template <class T>
List <T>::~List()
{
  iterator it = begin();
  while (!empty())
  {
    remove(it);
  }
}

/*
template <class T>
List<T>& List<T>::operator=(const List& li)
{
  if (&li != this)
  {
    while (!empty())
    {
      remove(begin());
    }
    for (iterator it=li.begin(); it!=li.end(); it++)
    {
      insert_before(end(),*it);
    }
  }
  return *this;
  }*/

template class List<int>;

////////////// LIST_ITERATOR CLASS DEFINITION //////////////

template <class T>
T& List_iterator<T>::operator*()
{
  return node->value;
}

template <class T>
List_iterator<T>& List_iterator<T>::operator++()
{
  node = node->succ;
  return *this;
}

template <class T>
List_iterator<T> List_iterator<T>::operator++(int)
{
  List_iterator orig = *this;
  ++(*this);
  return orig;
}


template <class T>
List_iterator<T>& List_iterator<T>::operator--()
{
  node = node->pred;
  return *this;
}

template <class T>
List_iterator<T> List_iterator<T>::operator--(int)
{
  List_iterator orig = *this;
  --(*this);
  return orig;
}

template<class T>
bool List_iterator<T>::operator==(const iterator& lit) const
{
    return (this->node == lit.node);
}

template <class T>
bool List_iterator<T>::operator!=(const iterator& lit) const
{
    return !(this->node == lit.node);
}

template class List_iterator<int>;

File: chapter4/list/Java/List.java
/*
 * Chapter 4
 * Paragraph 1
 *
 * List data structure implementation
 */

public class List<T> {
  private List pred;
  private List succ;
  private T value;

  public List()
  {
    this.pred = this;
    this.succ = this;
  }

  public boolean empty()
  {
    return (this.pred == this.succ) && (this.succ == this);
  }

  public List head()
  {
    return this.succ;
  }

  public List tail()
  {
    return this.pred;
  }

  public boolean finished(List p)
  {
    return (p == this);
  }

  public List next(List p)
  {
    return p.succ;
  }

  public List prev(List p)
  {
    return p.pred;
  }

  public T read()
  {
    return this.value;
  }

  public void write(List p, T v)
  {
    p.value = v;
  }

  public List insert(List p, T v)
  {
    List t = new List<T>();
    t.value = v;
    t.pred = p.pred;
    t.pred.succ = t;
    t.succ = p;
    p.pred = t;
    return p;
  }

  public List remove(List p)
  {
    p.pred.succ = p.succ;
    p.succ.pred = p.pred;
    List t = p.succ;
    return t;
  }

  public static void main(String args[]) {
    List<Integer> l = new List<Integer>();

    for (int i=0; i<10; i++)
    {
      l = l.insert(l,i);
    }

    List<Integer> it = l.head();
    while (!l.finished(it))
    {
      l.insert(it,15);
      l.write(it,20);
      it = l.next(it);
    }

    for (it=l.head(); !l.finished(it); it=l.next(it))
    {
      System.out.println(it.read());
    }
    System.out.println("\n");
  }
}